Add callback log througput

HDThang opened the pull request:
# Background
Trainer log loss and learning rate each logging_step in model.state. Throughput only calculate when finished the training process (`max_samples`/`run_time`). 
To calculate the throughput, I add a custom callback to trainer.
# What does this PR do?
Create callback to calculate throughput each log interval and save to model.state

HDThang committed:
add mlflow to train script for these tasks: token-cls, audio-cls, semantic-segmentation, and image-pretraining

HDThang committed:
-create call back calcualte througput each log inteval and save to model.state

dohyung-kim-moreh commented on examples/pytorch/token-classification/run_ner.py:
```diff
@@ -202,6 +209,30 @@ def __post_init__(self):
                 assert extension in ["csv", "json"], "`validation_file` should be a csv or a json file."
         self.task_name = self.task_name.lower()
 
+class MyCallback(TrainerCallback):
```
How about naming `TBTrainerCallback` or something else instead of `MyCallback` because `MyCallback` does not a specific meaning.

HDThang replied:
Sure, let's me change the callback name.

dohyung-kim-moreh commented on examples/pytorch/token-classification/run_ner.py:
```diff
@@ -202,6 +209,30 @@ def __post_init__(self):
+            logging_step_runtime = time.time() - self.start_time
+            num_samples = args.per_device_train_batch_size * args.logging_steps
+            throughput = num_samples / logging_step_runtime
+            state.log_history[-1]["throughput"] = throughput
```
While you're at it, why don't you work on printing out loss, lr, and evaluation metrics as well?

HDThang replied:
There is already print all the train and eval metrics at the end of the training, so I don't thing this necessary.

dohyung-kim-moreh replied:
But, loss and lr also should be printed out at every log interval. That's written in [Benchmark Code Guideline](https://www.notion.so/Benchmark-Code-Guideline-1c544796906a4ccea337ae289dab5b68?pvs=4).

dohyung-kim-moreh replied:
And evaluation metrics are expected to be logged at the evaluation phase.

dohyung-kim-moreh commented:
Could you upload the result of `tests/test_benchmkar.py` for some models in `transformers`?

HDThang commented:
@dohyung-kim-moreh I get 1 failed, and 5 passed with the old code

FAILED tests/test_benchmark.py::TestModelRunner::test_model[transformers-model_config0-benchmark_config0-env_config0] - AssertionError: Throughput should be logged via mlflow with step attribute.
 )

dohyung-kim-moreh commented:
Yes, throughput should be logged with `step` attribute.

HDThang committed:
-save metrics to mlflow and print them each log interval

HDThang commented:
@dohyung-kim-moreh  I've updated the code as per your suggestions: changed the callback name, printed out the metrics at each logging interval, and saved the metrics to MLflow

HDThang committed:
add mlflow and callback to summarization task

dohyung-kim-moreh commented on examples/pytorch/audio-classification/run_audio_classification.py:
```diff
@@ -391,6 +430,11 @@ def compute_metrics(eval_pred):
         train_result = trainer.train(resume_from_checkpoint=checkpoint)
         trainer.save_model()
+        for metric_dict in trainer.state.log_history:
+            if 'loss' in metric_dict:
```
It seems to call mlflow API after training is done. 
We should not assume a pytorch process is normally done. 
Any metrics won't be logged in the case that the process is terminated.

In this view, my suggestion is to log metrics in `TBTrainerCallback `.  Do you have any idea?

HDThang replied:
As I mention before, trainer log these metrics each log interval in trainer state. So those line of code just get the logged values in model states and save to mlflow. This way, we get the results as list with each metrics value associate with one log interval. Using `TBTrainerCallback `  to log these metrics every interval is trickier b/c we can get the result as list.

HDThang replied:
As I known, mlflow.log_metric take a value or a list for a key word. I don't know if it can append the new value to the same key word each interval.
Update: I find a way to make it work. The next commit will be push soon.

HDThang committed:
fix callback to log metrics each interval

yeomni commented:
LGTM with the latest commit

yeomni reviewed (approved):
Please check if the MLFlow log is properly collected after merging this one.

HDThang committed:
modify callback to log metric each interval for task: image-pretrain, token-cls, audio-cls and segmentation

HDThang committed:
add callback and mlflow api for text-cls, multiple-choice, language modeling and image-cls

HDThang committed:
add callback and mflow to train script to question answering, translation and summarization tasks
